OSOE-62: Include Lombiq.BaseTheme⭐

sarahelsaig opened the pull request:
OSOE-62
- https://github.com/Lombiq/Helpful-Extensions/pull/66
- https://github.com/Lombiq/Helpful-Libraries/pull/133
- https://github.com/Lombiq/Orchard-Base-Theme/pull/15
- https://github.com/Lombiq/Orchard-Data-Tables/pull/95
- https://github.com/Lombiq/Orchard-Training-Demo-Module/pull/77
- https://github.com/Lombiq/UI-Testing-Toolbox/pull/155
- https://github.com/Lombiq/Hosting-Tenants/pull/14
- https://github.com/Lombiq/Orchard-Azure-Application-Insights/pull/32
- https://github.com/Lombiq/Orchard-Chart.js/pull/20
- https://github.com/Lombiq/Orchard-JSON-Editor/pull/17
- https://github.com/Lombiq/Orchard-Privacy/pull/17
- https://github.com/Lombiq/Orchard-Vue.js/pull/45

sarahelsaig committed:
Add submodule.

sarahelsaig committed:
Add project to solution.

sarahelsaig committed:
Move stuff into subdirectory.

sarahelsaig committed:
Add sample project and path changes.

sarahelsaig committed:
Add sample theme to the recipe.

sarahelsaig committed:
Submodules and various fixes.

sarahelsaig committed:
Add navigation providers.

sarahelsaig committed:
Disambiguate async actions.

sarahelsaig committed:
Merge branch 'dev' into issue/OSOE-62

# Conflicts:
#	src/Libraries/Lombiq.HelpfulLibraries

sarahelsaig committed:
Merge remote-tracking branch 'origin/dev' into issue/OSOE-62

# Conflicts:
#	src/Libraries/Lombiq.HelpfulLibraries
#	src/Lombiq.OSOCE.Web/Lombiq.OSOCE.Web.csproj

sarahelsaig committed:
submodule

sarahelsaig committed:
Merge remote-tracking branch 'origin/dev' into issue/OSOE-62

# Conflicts:
#	src/Libraries/Lombiq.HelpfulLibraries

sarahelsaig committed:
Create testing docker image + readme.

sarahelsaig committed:
Add missing package.

sarahelsaig committed:
Merge remote-tracking branch 'origin/issue/OSOE-60' into issue/OSOE-62

# Conflicts:
#	src/Libraries/Lombiq.HelpfulLibraries
#	src/Modules/Lombiq.HelpfulExtensions

sarahelsaig committed:
Project update and cleanup.

sarahelsaig committed:
Move navigation into their respective projects.

sarahelsaig committed:
Add main menu navigation provider.

sarahelsaig committed:
Merge remote-tracking branch 'origin/dev' into issue/OSOE-62

# Conflicts:
#	src/Libraries/Lombiq.HelpfulLibraries
#	src/Modules/Lombiq.DataTables
#	src/Modules/Lombiq.OSOCE.Samples/Startup.cs
#	src/Modules/Lombiq.TrainingDemo

sarahelsaig committed:
We don't need this here.

sarahelsaig committed:
Merge remote-tracking branch 'origin/dev' into issue/OSOE-62

# Conflicts:
#	Lombiq.OSOCE.sln
#	src/Modules/Lombiq.TrainingDemo

sarahelsaig committed:
Merge branch 'dev' into issue/OSOE-62

# Conflicts:
#	Lombiq.OSOCE.sln
#	src/Libraries/Lombiq.HelpfulLibraries
#	src/Modules/Lombiq.DataTables
#	src/Modules/Lombiq.TrainingDemo

sarahelsaig committed:
Post-merge code cleanup.

sarahelsaig committed:
Make the menu widget recursive & bugfix TypedRoute.

sarahelsaig committed:
Add missing package.json file.

sarahelsaig committed:
Merge remote-tracking branch 'origin/dev' into issue/OSOE-62

# Conflicts:
#	src/Libraries/Lombiq.HelpfulLibraries

sarahelsaig committed:
submodules

sarahelsaig committed:
submodules

sarahelsaig committed:
Merge branch 'dev' into issue/OSOE-62

sarahelsaig committed:
Merge remote-tracking branch 'origin/dev' into issue/OSOE-62

# Conflicts:
#	src/Modules/Lombiq.DataTables
#	src/Modules/Lombiq.HelpfulExtensions

sarahelsaig committed:
Code styling.

sarahelsaig committed:
I really don't have the slightest idea.

sarahelsaig committed:
Fix sample menu.

sarahelsaig committed:
Menu and zone improvements.

sarahelsaig committed:
Merge remote-tracking branch 'origin/dev' into issue/OSOE-62

# Conflicts:
#	src/Modules/Lombiq.HelpfulExtensions

sarahelsaig committed:
submodules

sarahelsaig committed:
Code styling.

sarahelsaig committed:
Merge remote-tracking branch 'origin/dev' into issue/OSOE-62

# Conflicts:
#	src/Libraries/Lombiq.HelpfulLibraries
#	src/Modules/Lombiq.HelpfulExtensions

sarahelsaig committed:
Improve navigation

sarahelsaig committed:
submodules

sarahelsaig committed:
Fix HTML validation.

sarahelsaig committed:
Fix favicon related HTML validation.

sarahelsaig committed:
Fix regressions.

sarahelsaig committed:
Code styling.

sarahelsaig committed:
Fix test?

sarahelsaig committed:
Add tests.

sarahelsaig committed:
submodule

sarahelsaig committed:
Fix test.

Piedone commented on src/Lombiq.OSOCE.Web/Program.cs:
```diff
@@ -8,6 +9,9 @@
 
 var configuration = builder.Configuration;
 
+// I don't know why this is necessary.
```
Nevertheless, some comment on why you've added it is necessary.

sarahelsaig replied:
It's because otherwise I get this right after setup:
![image](https://user-images.githubusercontent.com/4007293/163898338-e8ae71c0-ba3b-40e9-9688-ec9b72a45d7d.png)
Strangely it only happens when you run Lombiq.OSOCE.Web directly, but not from the UI tests. Could you check if you get the same problem too if you comment out the line?

Piedone replied:
I get the same with the TEST recipe (during setup, if the site is already set up then it works without this line). This looks like an issue with OrchardCore.Roles or Users not being properly enabled.

sarahelsaig replied:
Yeah, I meant that it works with the test recipe in UI tests but not if you manually start it up. Turns out the extra dependency for "OrchardCore.Users" in "Lombiq.Tests.UI.Shortcuts" fixed it for UI testing. It's good now.

Piedone replied:
This needs to be done for the NuGet solution too.

sarahelsaig replied:
On second thought adding them to the setup recipe is a bit fragile. Instead, I have moved the `"OrchardCore.*"` features from the TrainingDemo recipe's feature step into the TrainingDemo's manifest file as dependencies. I think that's more correct, right? Then the OSOCE setup recipe doesn't need them either. I have added a comment about it in the _TrainingDemo.recipe.json_ too. I have checked and there aren't any other such features in any of the other recipes so I think we are in the clear.

This makes it self-contained so I don't think there there is anything NuGet specific to be done.

Piedone replied:
Yeah, in the manifest it's better.

Piedone commented on src/Lombiq.OSOCE.Web/Recipes/Lombiq.OSOCE.Tests.recipe.json:
```diff
@@ -89,9 +90,18 @@
-      // Overriding TheBlogTheme because that uses the PageTitle shape for <title> and thus Shape Tracing Helpful
-      // Extensions would mess it up.
-      "site": "TheTheme"
+      // Using a custom theme that relies on "Lombiq.BaseTheme". Previously we used "TheTheme" because the dafault
```
```suggestion
      // Using a custom theme that relies on "Lombiq.BaseTheme". Previously we used "TheTheme" because the default
```

Piedone commented on src/Modules/Lombiq.OSOCE.Samples/Startup.cs:
```diff
@@ -1,7 +1,12 @@
+    public override void ConfigureServices(IServiceCollection services)
+    {
+        // Nothing here yet.
+    }
```
Why did you add this?

Piedone commented on src/Modules/Lombiq.OSOCE.Samples/Lombiq.OSOCE.Samples.csproj:
```diff
@@ -23,7 +23,9 @@
   <ItemGroup>
-    <ProjectReference Include="..\..\Libraries\Lombiq.HelpfulLibraries\Lombiq.HelpfulLibraries\Lombiq.HelpfulLibraries.csproj" />
+    <ProjectReference Include="..\..\Libraries\Lombiq.HelpfulLibraries\Lombiq.HelpfulLibraries.Samples\Lombiq.HelpfulLibraries.Samples.csproj" />
+    <ProjectReference Include="..\..\Themes\Lombiq.BaseTheme\Lombiq.BaseTheme.Samples\Lombiq.BaseTheme.Samples.csproj" />
+    <ProjectReference Include="..\Lombiq.DataTables\Lombiq.DataTables.Samples\Lombiq.DataTables.Samples.csproj" />
```
Why are these necessary?

sarahelsaig replied:
I forgot to revert these.

Piedone commented on src/Lombiq.OSOCE.Web/Recipes/Lombiq.OSOCE.Tests.recipe.json:
```diff
@@ -18,6 +18,7 @@
       "disable": [],
       "enable": [
         "Lombiq.AuditTrailExtensions",
+        "Lombiq.BaseTheme.Samples",
```
Wow, this is nice!
![image](https://user-images.githubusercontent.com/1976647/163877356-16dca103-b421-44cd-9c71-317d7aa95f1a.png)

Add a note to the root Readme about adding a menu provider for this too.

Piedone commented on test/Lombiq.OSOCE.Tests.UI/Tests/BehaviorThemeTests.cs:
```diff
@@ -0,0 +1,27 @@
+
+namespace Lombiq.OSOCE.Tests.UI.Tests;
+
+public class BehaviorThemeTests : UITestBase
```
```suggestion
public class BehaviorBaseThemeTests : UITestBase
```

Piedone commented on src/Lombiq.OSOCE.Web/Recipes/Lombiq.OSOCE.Tests.recipe.json:
```diff
@@ -89,9 +90,18 @@
-      "site": "TheTheme"
+      // Using a custom theme that relies on "Lombiq.BaseTheme". Previously we used "TheTheme" because the dafault
+      // "TheBlogTheme" uses the PageTitle shape for <title> and Shape Tracing Helpful Extensions would mess it up.
+      "site": "Lombiq.BaseTheme.Samples"
```
`FrontendPagesShoudlBeAccessbile` can be adjusted.

sarahelsaig replied:
Adjusted. Unfortunately we still have to put up with some BS from the footer widget so only 2 of the 4 rules can be restored.
[FrontendPagesShouldBeAccessible.pdf](https://github.com/Lombiq/Open-Source-Orchard-Core-Extensions/files/8509000/FrontendPagesShouldBeAccessible.pdf)

Piedone replied:
We can just fix these, no?

sarahelsaig replied:
Not without replacing the blog recipe. It should be fixed upstream in OC.

Piedone replied:
OK, so that's about Blog. We did fix some of these (in the upcoming 1.4) I think.

Piedone replied:
I don't see any changes in `FrontendPagesShouldBeAccessible`.

Piedone replied:
Now I do!

Piedone commented on Lombiq.OSOCE.sln:
```diff
@@ -111,6 +111,16 @@ Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Lombiq.HelpfulLibraries.Orc
 EndProject
 Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Lombiq.HelpfulLibraries.OrchardCore.Testing", "src\Libraries\Lombiq.HelpfulLibraries\Lombiq.HelpfulLibraries.OrchardCore.Testing\Lombiq.HelpfulLibraries.OrchardCore.Testing.csproj", "{95D73C6A-B81E-44A3-9376-9CFEFDE3B314}"
 EndProject
+Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "Lombiq.BaseTheme", "src\Themes\Lombiq.BaseTheme\Lombiq.BaseTheme\Lombiq.BaseTheme.csproj", "{F482D324-1F7C-4D2A-8124-F456ABE58644}"
```
Do https://github.com/Lombiq/Open-Source-Orchard-Core-Extensions#:~:text=If%20an%20extension%20is%20added%2C%20removed%20or%20significantly%20updated%20in%20this%20project%2C%20then%20add%2C%20remove%20or%20update%20its%20description%20under%20the%20%22Included%20Projects%22%20section%20of%20this%20Readme.

sarahelsaig committed:
Update src/Lombiq.OSOCE.Web/Recipes/Lombiq.OSOCE.Tests.recipe.json

Co-authored-by: Zoltán Lehóczky <zoltan.lehoczky@lombiq.com>

sarahelsaig committed:
fixes

sarahelsaig committed:
Reset Lombiq.OSOCE.Samples to dev.

sarahelsaig committed:
Merge branch 'issue/OSOE-62' of https://github.com/Lombiq/Open-Source-Orchard-Core-Extensions into issue/OSOE-62

sarahelsaig committed:
Rename BehaviorThemeTests to BehaviorBaseThemeTests

sarahelsaig committed:
Update Included Projects.

sarahelsaig committed:
Elaborate.

sarahelsaig committed:
Adjust FrontendPagesShouldBeAccessible

sarahelsaig committed:
Added a note to the readme about adding main menu navigation providers.

sarahelsaig committed:
Clarify ForEach extension documentation.

sarahelsaig committed:
Rename Licence.md to License.md.

sarahelsaig committed:
Fix missing dependency in setup recipe.

sarahelsaig committed:
submodules

sarahelsaig committed:
Merge remote-tracking branch 'origin/dev' into issue/OSOE-62

# Conflicts:
#	src/Libraries/Lombiq.HelpfulLibraries
#	src/Modules/Lombiq.DataTables
#	src/Modules/Lombiq.HelpfulExtensions
#	src/Modules/Lombiq.TrainingDemo
#	test/Lombiq.UITestingToolbox

sarahelsaig committed:
Add nuget to samples as well.

sarahelsaig committed:
prerelease stuff

sarahelsaig committed:
Use relative path in _mixins.scss.

Piedone commented on Readme.md:
```diff
@@ -45,6 +45,10 @@ This also serves as an example of an ASP.NET Core web app using Orchard from NuG
 
+- Themes
+  - [Lombiq.BaseTheme](https://github.com/Lombiq/Orchard-Base-Theme/tree/dev): This theme contains infrastructure for custom Bootstrap 5 themes with standardized zones and built-in front end menu display.
+  - [Lombiq.BaseTheme.Samples](https://github.com/Lombiq/Orchard-Base-Theme/tree/dev/Lombiq.BaseTheme.Samples): A sample theme that demonstrates the features of the [Lombiq.BaseTheme](https://github.com/Lombiq/Orchard-Base-Theme/tree/dev).
```
Add a corresponding entry to "Samples and Recipes" too.

sarahelsaig committed:
submodules

sarahelsaig committed:
Merge remote-tracking branch 'origin/dev' into issue/OSOE-62

# Conflicts:
#	Lombiq.OSOCE.sln
#	src/Lombiq.OSOCE.Web/Lombiq.OSOCE.Web.csproj
#	src/Modules/Lombiq.DataTables
#	src/Modules/Lombiq.HelpfulExtensions
#	src/Modules/Lombiq.TrainingDemo

sarahelsaig committed:
renormalize

sarahelsaig committed:
Added to "Samples and Recipes"

sarahelsaig committed:
Fix TrainingDemo feature dependencies.

sarahelsaig committed:
Merge remote-tracking branch 'origin/dev' into issue/OSOE-62

# Conflicts:
#	test/Lombiq.UITestingToolbox

sarahelsaig committed:
Add analyzer condition to don't include in nuget.

sarahelsaig committed:
Add library version variables.

sarahelsaig committed:
Use current package version numbers.

sarahelsaig committed:
Update nuget project.

sarahelsaig committed:
Merge remote-tracking branch 'origin/dev' into issue/OSOE-62

# Conflicts:
#	src/Libraries/Lombiq.HelpfulLibraries
#	src/Modules/Lombiq.TrainingDemo

sarahelsaig committed:
Update NuGet versions.

sarahelsaig committed:
submodules

sarahelsaig committed:
Submodules.

sarahelsaig committed:
Merge remote-tracking branch 'origin/dev' into issue/OSOE-62

# Conflicts:
#	Directory.Build.props

sarahelsaig committed:
Simplify Windows HOME location.

sarahelsaig committed:
test and probe 1

sarahelsaig committed:
test and probe 2

sarahelsaig committed:
test and probe 2

sarahelsaig committed:
test and probe 4

sarahelsaig committed:
Remove unnecessary console.logs.

sarahelsaig committed:
Comment out includePaths as it wouldn't work with ProjectReference.

sarahelsaig committed:
Revert nuget versioning.

Piedone commented on Readme.md:
```diff
@@ -73,6 +77,9 @@ You can activate various sample content in the site:
   - Go to Features in the admin dashboard and select "Lombiq Data Tables - Samples".
   - Go to Recipes in the admin dashboard and select "Lombiq Data Tables - Sample Content - Employee".
+- [Lombiq Base Theme for Orchard Core](https://github.com/Lombiq/Orchard-Base-Theme/tree/dev):
+  - The "TEST: Basic Orchard Features" setup recipe automatically sets it up. If not using it, run the "Lombiq Orchard Core Base Theme - Layers and Zones" recipe, and then enable the theme in Admin > Design > Themes.
```
```suggestion
  - The "TEST: Basic Orchard Features" setup recipe automatically sets it up. If not using it, run the "Lombiq Orchard Core Base Theme - Layers and Zones" recipe, and then enable the theme in Admin → Design → Themes.
```

Piedone commented on Readme.md:
```diff
@@ -87,6 +94,7 @@ When adding a new extension, or significant new features to existing extensions,
 - For user-facing features add a recipe to it, demonstrating its usage with sample data. In that case, refer to it in the above section.
 - If no data is needed or if the feature is more library-like, add a sample project (or in addition to the recipe). Put this project into the root of the submodule, so to have the main project's and sample project's folders side by side.
 - Add lower-level unit/integration tests as necessary with the [Lombiq Testing Toolbox for Orchard Core](https://github.com/Lombiq/Testing-Toolbox/).
+- If the sample project includes MVC actions, create a service that inherits from `MainMenuNavigationProviderBase` and adds front-end main menu items. The top level menu item should have the project's shortened name and the submenu items the individual actions. If you have several controllers, use separators and labels as you can see in the [TrainingDemoNavigationProvider.cs](src/Modules/Lombiq.T ...
```
I'd perhaps link to the GitHub page instead, though during local development this is cool too.

sarahelsaig replied:
Apparently this doesn't work when clicking from GitHub so it's a no-go. I have replaced it with an absolute URL, though that now points to dev like you said in the other PR so it won't work until the Lombiq.TrainingDemo PR is merged.

Piedone replied:
Yeah, an absolute URL is what I meant as well.

Piedone commented on NuGetTest/Directory.Build.props:
```diff
@@ -0,0 +1,5 @@
+<Project>
+  <PropertyGroup>
+    <NuGetBuild>true</NuGetBuild>
```
This solution shouldn't permanently target this. In fact, it should never, since it is not published to NuGet.

sarahelsaig replied:
Isn't this solution supposed to simulate a NuGet build environment?

Piedone replied:
No, it's supposed to be a testbed of using the NuGet packages. `NuGetBuild` is when a NuGet package is being created and published.

sarahelsaig replied:
Okay, then changed it to NuGetTest and used that in _/Directory.Build.props_ too.

Piedone commented on NuGetTest/src/Lombiq.OSOCE.NuGet.Web/Lombiq.OSOCE.NuGet.Web.csproj:
```diff
@@ -10,21 +10,23 @@
-    <PackageReference Include="Lombiq.Tests.UI.Shortcuts" Version="2.0.2" />
-    <PackageReference Include="Lombiq.UIKit" Version="2.0.0" />
+    <PackageReference Include="Lombiq.AuditTrailExtensions" Version="2.*" />
+    <PackageReference Include="Lombiq.BaseTheme.Samples" Version="1.0.0-alpha.8" />
```
Why no floating version here too?

sarahelsaig replied:
doesn't work with pre-release
https://docs.microsoft.com/en-us/nuget/concepts/package-versioning#:~:text=floating%20versions%20do%20not%20resolve%20prerelease%20versions%20unless%20opted%20into

Piedone replied:
OK then.

Piedone commented on NuGetTest/src/Modules/Lombiq.OSOCE.NuGet.TestModule/Controllers/AnalyzersTestController.cs:
```diff
@@ -8,9 +8,9 @@ public class AnalyzersTestController : Controller
-    // Since the Analyzers NuGet package is added, you can see that it works like this: Disable the IDE0021 rule in the
-    // root Analyzers OrchardCore.ruleset file, then change this constructor to use a block body. You should get an
-    // analyzer violation since the expression-bodied rule is still enforced from the package
+    // Since the Analyzers NuGet package is added, you can see that it works like this: Change this constructor to use a
+    // block body. You should get an analyzer violation since the expression-bodied rule is still enforced from the
+    // package.
```
But it's also enforced from the submodule.

sarahelsaig replied:
This project should only be verified by the NuGet package of Lombiq.Analyzers. That's why:
- https://github.com/Lombiq/Open-Source-Orchard-Core-Extensions/pull/54/files/64602ff8917c799ae8906af78b458968047cbc73..1f6f5366d3a6e803946012ce2e45955b59f15be1#diff-b994a2e9f5cc46380b753ba286d391de4efd18431ae8ab574fe1dd45eb15d62bR3
- https://github.com/Lombiq/Open-Source-Orchard-Core-Extensions/blob/issue/OSOE-62/Directory.Build.props#L13

Piedone replied:
I see, good.

Piedone commented on NuGetTest/src/Modules/Lombiq.OSOCE.NuGet.TestModule/Lombiq.OSOCE.NuGet.TestModule.csproj:
```diff
@@ -16,8 +16,8 @@
   <ItemGroup>
-    <PackageReference Include="Lombiq.Analyzers" Version="2.0.0" />
-    <PackageReference Include="Lombiq.HelpfulLibraries" Version="4.0.0" />
+    <PackageReference Include="Lombiq.Analyzers" Version="2.0.2" />
+    <PackageReference Include="Lombiq.HelpfulLibraries" Version="4.1.0" />
```
Why no floating versions here and the other projects too?

sarahelsaig replied:
From what I understand they are kinda unnecessary in libraries because NuGet automatically does package upgrade in the dependency tree as needed.

sarahelsaig committed:
Make the "see here" links be part of a sentence too.

sarahelsaig committed:
Update Readme.md

Co-authored-by: Zoltán Lehóczky <zoltan.lehoczky@lombiq.com>

sarahelsaig committed:
submodule update

sarahelsaig committed:
Replace relative link that doesn't work on GitHub

because it goes into a submodule and apparently the site doesn't resolve it as if it was a symlink.

sarahelsaig committed:
Clarify prerelease comment.

sarahelsaig committed:
Use NuGetTestbed variable instead.

sarahelsaig committed:
Add menu label styling.

Piedone committed:
Updating submodules merged to dev

Piedone committed:
Updating version numbers
